repair/repair.cc

```diff
@@ -1540,3 +1540,109 @@ future<> sync_data_with_repair(seastar::sharded<database>& db,
         });
     });
 }
+
+future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, std::unordered_set<dht::token> tokens) {
+    using inet_address = gms::inet_address;
+    return seastar::async([&db, tm = std::move(tm), tokens = std::move(tokens)] () mutable {
+        auto keyspaces = db.local().get_non_system_keyspaces();
+        rlogger.info("bootstrap_with_repair: started with keyspaces={}", keyspaces);
+        auto myip = utils::fb_utilities::get_broadcast_address();
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                rlogger.info("bootstrap_with_repair: keyspace={} does not exist any more, ignoring it", keyspace_name);
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            dht::token_range_vector desired_ranges = strat.get_pending_address_ranges(tm, tokens, myip);
+
+            //Active ranges
+            auto metadata_clone = tm.clone_only_token_map();
+            auto range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Pending ranges
+            metadata_clone.update_normal_tokens(tokens, myip);
+            auto pending_range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Collects the source that will have its range moved to the new node
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+
+            rlogger.info("bootstrap_with_repair: started with keyspace={}, nr_ranges={}", keyspace_name, desired_ranges.size());
+            for (auto& desired_range : desired_ranges) {
+                for (auto& x : range_addresses) {
+                    const range<dht::token>& src_range = x.first;
+                    seastar::thread::maybe_yield();
+                    if (src_range.contains(desired_range, dht::tri_compare)) {
```

Comment:
Should this be .intersects()? Or will the ranges exactly match?

Reply:
It is `contains` because the new node range can only be the subset of the old range during bootstrap, e.g., old range is (1, 100], the new node now owns (1, 50].

Reply:
Right.